devtools/x/src/lint/guppy.rs

```diff
@@ -237,3 +241,54 @@ impl PackageLinter for IrrelevantBuildDeps {
         Ok(RunStatus::Executed)
     }
 }
+
+/// Ensure that packages within the workspace only depend on one version of a third-party crate.
+#[derive(Debug)]
+pub struct DirectDepDups;
+
+impl Linter for DirectDepDups {
+    fn name(&self) -> &'static str {
+        "direct-dep-dups"
+    }
+}
+
+impl ProjectLinter for DirectDepDups {
+    fn run<'l>(
+        &self,
+        ctx: &ProjectContext<'l>,
+        out: &mut LintFormatter<'l, '_>,
+    ) -> Result<RunStatus<'l>> {
+        let package_graph = ctx.package_graph()?;
+
+        // This is a map of direct deps by name -> version -> packages that depend on it.
+        let mut direct_deps: BTreeMap<&str, BTreeMap<&Version, Vec<&str>>> = BTreeMap::new();
+        package_graph.query_workspace().resolve_with_fn(|_, link| {
```

Comment:
Very happy I made `resolve_with_fn` a `FnMut` so it can be abused in this way without using a RefCell, hehe ;)